src/bluesearch/database/topic_rule.py

```diff
@@ -0,0 +1,125 @@
+# Blue Brain Search is a text mining toolbox focused on scientific use cases.
+#
+# Copyright (C) 2020  Blue Brain Project, EPFL.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program. If not, see <https://www.gnu.org/licenses/>.
+"""Implementation of the TopicRule data structure."""
+from __future__ import annotations
+
+import re
+from typing import Iterable, Optional
+
+import pydantic
+
+from bluesearch.database.article import ArticleSource
+from bluesearch.database.topic_info import TopicInfo
+
+
+class TopicRule(pydantic.BaseModel, arbitrary_types_allowed=True):
+    """The topic rule."""
+
+    # None always represent wildcards
+    level: Optional[str] = None  # "article" or "journal"
+    source: Optional[ArticleSource] = None  # "arxiv", ... , "pubmed"
+    pattern: Optional[re.Pattern] = None  # regex pattern to match
+
+    @pydantic.validator("level")
+    def check_level_value(cls, value):
+        """Check the level parameter value."""
+        if value is not None and value not in {"article", "journal"}:
+            raise ValueError(f"Unsupported level {value}")
+        return value
+
+    @pydantic.validator("source", pre=True)
+    def convert_source_to_article_source(cls, value) -> ArticleSource | None:
+        """Check the source parameter value."""
+        if value is not None:
+            try:
+                value = ArticleSource(value)
+            except ValueError:
+                raise ValueError(f"Unsupported source {value}") from None
+
+        return value
+
+    @pydantic.validator("pattern", pre=True)
+    def convert_pattern_to_re(cls, value) -> re.Pattern | None:
+        """Check the pattern parameter value."""
+        if value is not None:
+            try:
+                value = re.compile(value)
+            except re.error:
+                raise ValueError(f"Unsupported pattern {value}") from None
+
+        return value
```

Comment:
```suggestion
from typing import Iterable

from bluesearch.database.article import ArticleSource
from bluesearch.database.topic_info import TopicInfo


class TopicRule:
    """Rule for accepting/rejecting an article based on topic matching criteria.

    Parameters
    ----------
    level
        Level of the topic information to match, must be "article" or "journal".
        Passing `None` will match any level.
    source
        Article source, must be a valid ArticleSource (e.g. "arxiv", "pmc", ...).
        Passing `None` will match any source.
    pattern
        Regular expression for matching the topic names of a given article.
        Passing `None` will match the name of any topic.
    """

    def __init__(
        self,
        level: str | None = None,
        source: str | ArticleSource | None = None,
        pattern: str | re.Pattern | None = None,
    ):
        if level is not None and level not in {"article", "journal"}:
            raise ValueError(f"Unsupported level {level}.")
        self.level = level
        self.source = ArticleSource(source) if ArticleSource is not None else None
        self.pattern = re.compile(pattern) if pattern is not None else None
```

Reply:
Didn't commit the suggestion but copy-pasted by hand since I had to revert the pydantic commit first. All changes are here: https://github.com/BlueBrain/Search/pull/550/files/b141457d414cb35b833164ad1e63ef4fd2cc32cc..8f2e1960d33e6d24199849023341a349118abf11